### PR TITLE
[WIP] Batched drivers with spin moves

### PR DESCRIPTION
**Pull request by camelto2**
Please review the [developer documentation](https://github.com/QMCPACK/qmcpack/wiki/Development-workflow)
on the wiki of this project that contains help and requirements.

## Proposed changes
This PR adds the functionality needed to perform spin moves in the batched drivers, needed for SOC. Essentially, I added a new advanceWalkersWithSpin which is a copy of advanceWalkers and adds the required spin moves and WF calls. The selection of which advanceWalkers call is used is determined at the runQMCstep level after checking if the particleset leader is a spinor or not. Adding this new function was much cleaner than adding a bunch of if statements everywhere, and until we have a cleaner templated solution this seemed like the easiest path forward. Some documentation is added to the batched drivers since the spin_mass variable is added to QMCDriverInput

Part of #3712 

## What type(s) of changes does this code introduce?
_Delete the items that do not apply_

- New feature

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?
macOSX

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- No. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- Yes. Documentation has been added (if appropriate)

### CODE REVIEW

File: src/QMCDrivers/VMC/VMCBatched.cpp [hunk 1]
```diff
@@ -215,6 +215,216 @@ void VMCBatched::advanceWalkers(const StateForThread& sft,
...
+
+  timers.movepbyp_timer.start();
+  const int num_walkers = crowd.size();
+  // Note std::vector<bool> is not like the rest of stl.
+  std::vector<bool> moved(num_walkers, false);
+  constexpr RealType mhalf(-0.5);
+  const bool use_drift = sft.vmcdrv_input.get_use_drift();
+  std::vector<TrialWaveFunction::GradType> grads_now(num_walkers);
+  std::vector<TrialWaveFunction::GradType> grads_new(num_walkers);
+  std::vector<TrialWaveFunction::ComplexType> spingrads_now(num_walkers);
```
Comment (ye-luo) [inline src/QMCDrivers/VMC/VMCBatched.cpp:251]: Divergent

File: src/QMCDrivers/VMC/VMCBatched.cpp [hunk 2]
```diff
@@ -215,6 +215,216 @@ void VMCBatched::advanceWalkers(const StateForThread& sft,
...
+  // local list to handle accept/reject
+  std::vector<bool> isAccepted;
+  std::vector<std::reference_wrapper<TrialWaveFunction>> twf_accept_list, twf_reject_list;
+  isAccepted.reserve(num_walkers);
+
+  for (int sub_step = 0; sub_step < sft.qmcdrv_input.get_sub_steps(); sub_step++)
+  {
+    //This generates an entire steps worth of deltas.
+    step_context.nextDeltaRs(num_walkers * sft.population.get_num_particles());
+    step_context.nextDeltaSpins(num_walkers * sft.population.get_num_particles());
```
Comment (ye-luo) [inline src/QMCDrivers/VMC/VMCBatched.cpp:270]: Divergent

File: src/QMCDrivers/VMC/VMCBatched.cpp [hunk 3]
```diff
@@ -215,6 +215,216 @@ void VMCBatched::advanceWalkers(const StateForThread& sft,
...
+        auto delta_r_start    = step_context.deltaRsBegin() + iat * num_walkers;
+        auto delta_r_end      = delta_r_start + num_walkers;
+        auto delta_spin_start = step_context.deltaSpinsBegin() + iat * num_walkers;
+        auto delta_spin_end   = delta_spin_start + num_walkers;
+
+        if (use_drift)
+        {
+          twf_dispatcher.flex_evalGradWithSpin(walker_twfs, walker_elecs, iat, grads_now, spingrads_now);
+          sft.drift_modifier.getDrifts(tauovermass, grads_now, drifts);
+          sft.drift_modifier.getDrifts(spintauovermass, spingrads_now, spindrifts);
```
Comment (ye-luo) [inline src/QMCDrivers/VMC/VMCBatched.cpp:299]: Divergent

File: src/QMCDrivers/VMC/VMCBatched.cpp [hunk 4]
```diff
@@ -215,6 +215,216 @@ void VMCBatched::advanceWalkers(const StateForThread& sft,
...
+                         [sqrtspintau](const ParticleSet::Scalar_t& delta_spin) { return sqrtspintau * delta_spin; });
+        }
+
+        ps_dispatcher.flex_makeMoveWithSpin(walker_elecs, iat, drifts, spindrifts);
+
+        // This is inelegant
+        if (use_drift)
+        {
+          twf_dispatcher.flex_calcRatioGrad(walker_twfs, walker_elecs, iat, ratios, grads_new);
+          twf_dispatcher.flex_calcRatioGradWithSpin(walker_twfs, walker_elecs, iat, ratios, grads_new, spingrads_new);
```
Comment (ye-luo) [inline src/QMCDrivers/VMC/VMCBatched.cpp:325]: Divergent

Comment (ye-luo) [reply src/QMCDrivers/VMC/VMCBatched.cpp:325]: need both 324 and 325 lines?

File: src/QMCDrivers/VMC/VMCBatched.cpp [hunk 5]
```diff
@@ -215,6 +215,216 @@ void VMCBatched::advanceWalkers(const StateForThread& sft,
...
+                         [oneover2spintau](const ParticleSet::Scalar_t& spindrift, const RealType& lgb) {
+                           return lgb - oneover2spintau * spindrift * spindrift;
+                         });
+        }
+        else
+        {
+          twf_dispatcher.flex_calcRatio(walker_twfs, walker_elecs, iat, ratios);
+        }
+
+        std::transform(ratios.begin(), ratios.end(), prob.begin(), [](auto ratio) { return std::norm(ratio); });
```
Comment (ye-luo) [inline src/QMCDrivers/VMC/VMCBatched.cpp:357]: From this line and below, the code is spin agnostic. We need to think harder about an abstract to wrap the divergence between spin and non-spin cases.

**Review by ye-luo — CHANGES_REQUESTED** [review]

File: src/QMCDrivers/VMC/VMCBatched.cpp (hunk 4 above)
Comment (camelto2) [reply src/QMCDrivers/VMC/VMCBatched.cpp:325]: oops. good catch

File: src/QMCDrivers/VMC/VMCBatched.cpp (hunk 5 above)
Comment (camelto2) [reply src/QMCDrivers/VMC/VMCBatched.cpp:357]: same is true in the DMC batched.

File: src/QMCDrivers/VMC/VMCBatched.cpp (hunk 4 above)
Comment (camelto2) [reply src/QMCDrivers/VMC/VMCBatched.cpp:324]: fixed

**Comment by prckent** [comment]
(General comment, not intended to slow this down in any way) We have discussed this in the past, but to leave a record here - while the batched spin move drivers are worked on and merged, we should all think about how it might be possible to write a unified driver in the future that handles the non-spin move/spinor case as well. This is a matter of more than "simple" abstractions since the Monte Carlo algorithms are different, but they are not that different. One of the problems with the old CPU code is the proliferation of drivers, which adds risk, testing, & maintenance burden due to the proliferation of code paths. Would be good to avoid in the new code if practical.

**Comment by camelto2** [comment]
made this WIP while I work on an abstraction layer after talking with Ye and in line with Paul's comments above. Might be better to go ahead and get that sorted out sooner rather than later

**Pull request closed without merging**